### DISCUSSION TITLE
How to get structured output using pydantic defined model?

### DISCUSSION THREAD

**Discussion by SDAravind** [crewAIInc/crewAI#1436] [Q&A]
I tried creating pydantic base model and passed `Article` model to `Crew` class  but doesn't provide the structured output. What am I missing? 

Pydantic is None in the output.

Can you point me in right direction?  I have included the error in the end. I have no clue to address that error without a documentation. Looked into this github issue on LiteLLM [#4817](https://github.com/BerriAI/litellm/issues/4817)

FYI: I'm using Azure OpenAI - GPT-4O

```python
from crewai import Agent, LLM, Task, Crew, Process

from pydantic import BaseModel

class Article(BaseModel):
    title: str
    content: str

llm = LLM(
    model="YourModel",
    temperature=0.7,
    seed=42,
    base_url="https://YOURURL",
    api_key="LLM_API_KEY",
    api_version="LLM_API_VERSION",
    verbose=True,
)

researcher = Agent(
    role="Senior Researcher: {topic}",
    goal="Uncover groundbreaking technologies in AI: {topic}",
    backstory="Driven by curiosity, you explore and share the latest innovations.",
    verbose=True,
    llm=llm,
)


research_task = Task(
    description="Identify the next big trend in AI for {topic} with pros and cons.",
    expected_output="""A 3-paragraph article on emerging AI technologies with eye catching title and content.
    """,
    agent=researcher,
    # output_json=Article,
    # output_pydantic=Article,
    output_file="output/article.md",
)

crew = Crew(
    agents=[researcher],
    tasks=[research_task],
    process=Process.sequential,
    verbose=True,
)
```

### Output: I get the output but not the pydantic structured output.
> CrewOutput(raw="**AI-Driven Finance: Unveiling the Next Big Revolution**\n\nIn the ever-evolving landscape of finance, artificial intelligence (AI) is set to drive the next big trend: AI-powered hyper-personalized financial services. As the financial sector increasingly adopts AI technologies, institutions can offer services tailored to the individual needs and behaviors of each client. This level of personalization promises enhanced customer satisfaction, as financial advice, products, and offers can be precisely aligned with personal financial goals and risk appetites. Moreover, AI's ability to analyze vast amounts of data in real time allows for more accurate risk assessments, potentially increasing profitability while minimizing risk.\n\nDespite its promising potential, AI-driven finance comes with its own set of challenges. Privacy concerns are paramount, as hyper-personalized services require extensive data collection and analysis. Financial institutions must navigate the thin line between offering personalized experiences and maintaining stringent data protection measures. Additionally, while AI systems are generally proficient in handling structured data, they often struggle with the nuances of unstructured data, which can limit their effectiveness. Furthermore, implementing such sophisticated AI solutions can be costly, and the disparity in AI resource allocation might widen the gap between large institutions and smaller firms.\n\nLooking ahead, the integration of AI in finance represents a transformative opportunity, reshaping how financial services are delivered and consumed. As technology continues to advance, the focus on ethical AI usage will intensify, encouraging more transparent, fair, and equitable services. Financial institutions that can harness the power of AI responsibly will likely lead the charge in this revolutionary shift, setting new standards in an industry ripe for innovation. The key to success lies in balancing technological possibilities with ethical responsibility, ultimately fostering a financial environment that benefits all stakeholders involved.", pydantic=None, json_dict=None, tasks_output=[TaskOutput(description='Identify the next big trend in AI for finance with pros and cons.', name=None, expected_output='A 3-paragraph article on emerging AI technologies with eye catching title and content.\n    ', summary='Identify the next big trend in AI for finance with...', raw="**AI-Driven Finance: Unveiling the Next Big Revolution**\n\nIn the ever-evolving landscape of finance, artificial intelligence (AI) is set to drive the next big trend: AI-powered hyper-personalized financial services. As the financial sector increasingly adopts AI technologies, institutions can offer services tailored to the individual needs and behaviors of each client. This level of personalization promises enhanced customer satisfaction, as financial advice, products, and offers can be precisely aligned with personal financial goals and risk appetites. Moreover, AI's ability to analyze vast amounts of data in real time allows for more accurate risk assessments, potentially increasing profitability while minimizing risk.\n\nDespite its promising potential, AI-driven finance comes with its own set of challenges. Privacy concerns are paramount, as hyper-personalized services require extensive data collection and analysis. Financial institutions must navigate the thin line between offering personalized experiences and maintaining stringent data protection measures. Additionally, while AI systems are generally proficient in handling structured data, they often struggle with the nuances of unstructured data, which can limit their effectiveness. Furthermore, implementing such sophisticated AI solutions can be costly, and the disparity in AI resource allocation might widen the gap between large institutions and smaller firms.\n\nLooking ahead, the integration of AI in finance represents a transformative opportunity, reshaping how financial services are delivered and consumed. As technology continues to advance, the focus on ethical AI usage will intensify, encouraging more transparent, fair, and equitable services. Financial institutions that can harness the power of AI responsibly will likely lead the charge in this revolutionary shift, setting new standards in an industry ripe for innovation. The key to success lies in balancing technological possibilities with ethical responsibility, ultimately fostering a financial environment that benefits all stakeholders involved.", pydantic=None, json_dict=None, agent='Senior Researcher: finance', output_format=<OutputFormat.PYDANTIC: 'pydantic'>)], token_usage=UsageMetrics(total_tokens=552, prompt_tokens=196, completion_tokens=356, successful_requests=1))

### Error: I get this error
> Failed to convert text into a pydantic model due to the following error: litellm.APIError: AzureException APIError - argument of type 'NoneType' is not iterable Using raw output instead.

**Comment by niradar** [comment]
Same problem here. Didn't find the answer yet. Following the discussion

**Comment by NagarajVaidya** [comment]
Same problem here. I am clueless on these errors:

```
output_json mode:
Failed to convert text into JSON, error: litellm.APIError: AzureException APIError - argument of type 'NoneType' is not iterable. Using raw output instead.

output_pydantic mode:
Failed to convert text into a pydantic model due to the following error: litellm.APIError: AzureException APIError - argument of type 'NoneType' is not iterable Using raw output instead.
```

I am using Azure OpenAI - GPT-4o 2024-06-01

Below is my code for the Pydantic Model to get the structured output:

````python
from typing import List
from pydantic import BaseModel, Field

class InvoiceLineItemDetails(BaseModel):
    invoice_number: str = Field(..., description="Invoice Number available in the invoice")
    po_no: str = Field(..., description="Purchase Order Number against each line-item in the invoice")
    po_line_no: str = Field(..., description="Purchase Order Line Number associated with each Purchase Order")
    material_no: str  = Field(..., description="Material Number of each line-item in the invoice")
    quantity: str  = Field(..., description="Quantity of each line-item specified in the invoice")

class InvoiceData(BaseModel):
    invoice_data: List[InvoiceLineItemDetails] = Field(..., description="List of invoice line items details")
````

**Comment by NagarajVaidya** [comment]
Update:

I fixed my error by passing required keys in one of the task description and also corrected my pydantic model to match the same keys. 

I am getting expected results now! :D

I corrected my above pydantic model code as below

```python
from typing import List
from pydantic import BaseModel, Field

class InvoiceLineItemDetails(BaseModel):
    invoice_number: str = Field(..., description="Invoice Number available in the invoice")
    po_no: str = Field(..., description="Purchase Order Number against each line-item in the invoice")
    po_line: str = Field(..., description="Purchase Order Line Number associated with each Purchase Order")
    material_no: str  = Field(..., description="Material Number of each line-item in the invoice")
    quantity: str  = Field(..., description="Quantity of each line-item specified in the invoice")

class InvoiceData(BaseModel):
    invoice_data: List[InvoiceLineItemDetails] = Field(..., description="List of invoice line items details")
```

As you can see, I changed po_line_no to po_line because in one of my task description I was specifying this key as po_line.

Now, it is working as expected and getting the output in a structured way

![image](https://github.com/user-attachments/assets/d2ca9eba-6a4a-4909-8aba-316eb175e973)

**Comment by alvarolj23** [comment]
Hello, 
Same problem here, where you able to fix it?

**Comment by niradar** [comment]
No success here. I stopped using crewAI because of it and thought to check again later this month for a solution

**Comment by pip-install-skills** [comment]
any update on the output_pydantic parameter usage?

I was getting this error if I'm using pydantic classes


Traceback (most recent call last):
  File "D:\POC\test.py", line 54, in <module>
    task = Task(
           ^^^^^
  File "D:\POC\venv\Lib\site-packages\pydantic\main.py", line 212, in __init__
    validated_self = self.__pydantic_validator__.validate_python(data, self_instance=self)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pydantic_core._pydantic_core.ValidationError: 1 validation error for Task
expected_output
  Input should be a valid string [type=string_type, input_value=<class '__main__.PatientInfo'>, input_type=ModelMetaclass]
    For further information visit https://errors.pydantic.dev/2.9/v/string_type

**Reply by pip-install-skills** [reply]
this is my pydantic class declaration

```python
class DiseaseInfo(BaseModel):
    name: str = Field(..., description="Name of the disease")
    diagnosed_date: Optional[date] = Field(None, description="Date when the disease was diagnosed")
    severity: Optional[str] = Field(None, description="Severity level of the disease (e.g., mild, moderate, severe)")
    ongoing: bool = Field(..., description="Whether the disease is currently ongoing")

class ContactInfo(BaseModel):
    phone: Optional[str] = Field(None, description="Patient's contact phone number")
    email: Optional[str] = Field(None, description="Patient's contact email address")

class Address(BaseModel):
    street: str = Field(..., description="Street address")
    city: str = Field(..., description="City")
    state: str = Field(..., description="State")
    zip_code: Optional[str] = Field(None, description="ZIP or postal code")
    country: str = Field(..., description="Country")

class PatientInfo(BaseModel):
    id: str = Field(..., description="Unique identifier for the patient")
    name: str = Field(..., description="Patient's full name")
    age: int = Field(..., description="Patient's age")
    gender: Optional[str] = Field(None, description="Patient's gender")
    date_of_birth: Optional[date] = Field(None, description="Patient's date of birth")
    contact_info: Optional[ContactInfo] = Field(None, description="Patient's contact information")
    address: Optional[Address] = Field(None, description="Patient's address")
    diseases: List[DiseaseInfo] = Field(default_factory=list, description="List of diseases associated with the patient")
    notes: Optional[str] = Field(None, description="Additional notes about the patient")
```
